eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Pool.hs

```diff
@@ -172,15 +173,24 @@ instance Era era => DecCBOR (ShelleyPoolPredFailure era) where
       k -> invalidKey k
 
 poolDelegationTransition ::
-  forall era.
-  (ShelleyEraTxCert era, EraPParams era) =>
-  TransitionRule (ShelleyPOOL era)
+  forall (ledger :: Type -> Type) era.
+  ( ShelleyEraTxCert era
+  , EraPParams era
+  , Signal (ledger era) ~ TxCert era
+  , Environment (ledger era) ~ PoolEnv era
+  , State (ledger era) ~ PState era
+  , STS (ledger era)
+  , Event (ledger era) ~ PoolEvent era
+  , BaseM (ledger era) ~ ShelleyBase
+  , PredicateFailure (ledger era) ~ ShelleyPoolPredFailure era
+  ) =>
+  TransitionRule (ledger era)
 poolDelegationTransition = do
   TRC (PoolEnv slot pp, ps, c) <- judgmentContext
   let stpools = psStakePoolParams ps
   let pv = pp ^. ppProtocolVersionL
-  case c of
-    RegPoolTxCert poolParam -> do
+  case (getRegPoolTxCert c, getRetirePoolTxCert c) of
```

Comment:
@lehins created [patterns synonyms](https://github.com/input-output-hk/cardano-ledger/blob/master/eras/shelley/impl/src/Cardano/Ledger/Shelley/TxCert.hs#L173) that don't assume a concrete TxCert. You should use these instead.

Reply:
it shouldn't be necessary to change this function at all.

Reply:
oh, I see you are injecting the ledger rule into it, that makes sense